tests/test_plugin.py

```diff
@@ -261,3 +269,268 @@ def test_crud_for_navigation_plugin_when_versioning_disabled(self):
+            language=self.language, version__created_by=self.get_superuser()
+        )
+        placeholder = factories.PlaceholderFactory(source=page_content)
+        menu_content = factories.MenuContentWithVersionFactory()
+        child = factories.ChildMenuItemFactory(parent=menu_content.root)
+        grandchild = factories.ChildMenuItemFactory(parent=child)
+
+        # Patch the choices on the template field, so we don't get
+        # form validation errors
+        template_field = [
+            field for field in NavigationPlugin._meta.fields if field.name == "template"
+        ][0]
+        patched_choices = [
+            ("menu/menu.html", "Default"),
+            ("menu/menuismo.html", "Menuismo"),
+        ]
+        with patch.object(template_field, "choices", patched_choices):
+            # First add the plugin and assert
+            # The added plugin will have the template menu/menuismo.html
+            # and the menu from menu1
+            created_plugin = self._add_nav_plugin_and_assert(
+                placeholder, menu_content.menu, "menu/menuismo.html"
+            )
+
+            # # Now edit the plugin and assert
+            # # After editing the plugin will have the template menu/menu.html
+            # # and the menu from menu2
+            self._edit_nav_plugin_and_assert(
+                created_plugin, menu_content.menu, "menu/menu.html"
+            )
+
+        # Now publish the page content containing the plugin,
+        # so the page can be viewed
+        version = page_content.versions.get()
+        version.publish(self.get_superuser())
+
+        # asserting to check there is no cachekey exists
+        cache_key = CacheKey.objects.all().count()
+        self.assertEqual(cache_key, 0)
+        # And view the page
+        page_url = page_content.page.get_absolute_url()
+        response = self.client.get(page_url)
+
+        # rendering should generate cachekey object
+        cache_key = CacheKey.objects.all().count()
+        self.assertEqual(cache_key, 1)
+
+        self.assertEqual(response.status_code, 200)
+        self.assertIn(child.title, str(response.content))
+        self.assertIn(grandchild.title, str(response.content))
+
+        # create further menu items which covers draft state
+        child2 = factories.ChildMenuItemFactory(parent=menu_content.root)
+        # grandchild2
+        factories.ChildMenuItemFactory(parent=child2)
+
+        # MenuItem creation should be invalidated cache_key object
+        cache_key = CacheKey.objects.all().count()
+        self.assertEqual(cache_key, 0)
+
+        # Getting response again to fetch child2 and grandchild2 menu item
+        response = self.client.get(page_url)
+        self.assertEqual(response.status_code, 200)
+
+        # rendering should generate cachekey object
+        cache_key = CacheKey.objects.all().count()
+        self.assertEqual(cache_key, 1)
+
+    def test_menu_cache_invalidate_after_menucontent_publish(self):
```

Comment:
Why such complex tests for this? I don't understand why you need to create so much data (child-grandchild structures etc. which will be slower) when all you're asserting is the CacheKey? And of course the shorter a test the easier it is to read usually. So is there a reason for how you've chosen to do this?
I would think you achieve the same thing with something like:
```
def test_menu_cache_invalidated_on_publish(self):
    version = factories.MenuVersionFactory()
    self.assertEqual(CacheKey.objects.all().count(), 0)
    version.publish(self.get_superuser())
    self.assertEqual(CacheKey.objects.all().count(), 1)
```
Is that not so?

Reply:
`CacheKey` only generated once the menu is rendered on the frontend. The test is creating `Menucontent` with child objects then rendering it on the frontend and then checking cache is invalidated or not then I add another child item and checking it appears as well.